WalletWasabi/Transactions/TransactionStore.cs

```diff
@@ -0,0 +1,350 @@
+			}
+
+			return isRemoved;
+		}
+
+		public ISet<SmartTransaction> TryRemove(IEnumerable<uint256> hashes)
+		{
+			hashes ??= Enumerable.Empty<uint256>();
+			var removed = new HashSet<SmartTransaction>();
+
+			lock (TransactionsLock)
+			{
+				foreach (var hash in hashes)
+				{
+					if (Transactions.Remove(hash, out SmartTransaction stx))
+					{
+						removed.Add(stx);
+					}
+				}
+			}
+
+			if (removed.Any())
+			{
+				_ = TryRemoveFromFileAsync(removed.Select(x => x.GetHash()).ToArray());
+			}
+
+			return removed;
+		}
+
+		public bool TryGetTransaction(uint256 hash, out SmartTransaction sameStx)
+		{
+			lock (TransactionsLock)
+			{
+				return Transactions.TryGetValue(hash, out sameStx);
+			}
+		}
+
+		public ISet<SmartTransaction> GetTransactions()
+		{
+			lock (TransactionsLock)
+			{
+				return Transactions.Values.ToHashSet();
+			}
+		}
+
+		public ISet<uint256> GetTransactionHashes()
+		{
+			lock (TransactionsLock)
+			{
+				return Transactions.Keys.ToHashSet();
+			}
+		}
+
+		public bool IsEmpty()
+		{
+			lock (TransactionsLock)
+			{
+				return !Transactions.Any();
+			}
+		}
+
+		public bool Contains(uint256 hash)
+		{
+			lock (TransactionsLock)
+			{
+				return Transactions.ContainsKey(hash);
+			}
+		}
+
+		#region Serialization
+
+		private async Task SerializeAllTransactionsNoMutexAsync()
+		{
+			List<SmartTransaction> transactionsClone;
+			lock (TransactionsLock)
+			{
+				transactionsClone = Transactions.Values.ToList();
+			}
+
+			await TransactionsFileManager.WriteAllLinesAsync(transactionsClone.ToBlockchainOrderedLines()).ConfigureAwait(false);
+		}
+
+		private async Task TryAppendToFileAsync(params SmartTransaction[] transactions)
+			=> await TryAppendToFileAsync(transactions as IEnumerable<SmartTransaction>).ConfigureAwait(false);
+
+		private async Task TryAppendToFileAsync(IEnumerable<SmartTransaction> transactions)
+			=> await TryCommitToFileAsync(new Append(transactions)).ConfigureAwait(false);
+
+		private async Task TryRemoveFromFileAsync(params uint256[] transactionIds)
+			=> await TryRemoveFromFileAsync(transactionIds as IEnumerable<uint256>).ConfigureAwait(false);
+
+		private async Task TryRemoveFromFileAsync(IEnumerable<uint256> transactionIds)
+			=> await TryCommitToFileAsync(new Remove(transactionIds)).ConfigureAwait(false);
+
+		private List<ITxStoreOperation> Operations { get; } = new List<ITxStoreOperation>();
+		private object OperationsLock { get; } = new object();
+
+		private async Task TryCommitToFileAsync(ITxStoreOperation operation)
+		{
+			try
+			{
+				if (operation is null || operation.IsEmpty)
+				{
+					return;
+				}
+
+				// Make sure that only one call can continue.
+				lock (OperationsLock)
+				{
+					var isRunning = Operations.Any();
+					Operations.Add(operation);
+					if (isRunning)
+					{
+						return;
+					}
+				}
+
+				// Wait until the operation list calms down.
+				List<ITxStoreOperation> operationsToExecute;
```

Comment:
Can be an array.

Reply:
It can. Why would it be?

Reply:
If it won't change in the future and the Lenght is needed I usually use array. If the Length not needed and foreach-ed later I would go with IEnumerable to opt-out enumeration and copy at the end of the linq.

Reply:
Change it or left it like this on the other hand I will approve this.

Reply:
Changed to IEnumerable.